src/realm/sync/client.cpp

```diff
@@ -228,14 +228,8 @@ class SessionWrapper final : public util::AtomicRefCountBase, DB::CommitListener
     std::vector<WaitOperCompletionHandler> m_download_completion_handlers;
     std::vector<WaitOperCompletionHandler> m_sync_completion_handlers;
 
-    // `m_target_*load_mark` and `m_reached_*load_mark` are protected by
-    // `m_client.m_mutex`. `m_staged_*load_mark` must only be accessed by the
-    // event loop thread.
-    std::int_fast64_t m_target_upload_mark = 0, m_target_download_mark = 0;
-    std::int_fast64_t m_staged_upload_mark = 0, m_staged_download_mark = 0;
-    std::int_fast64_t m_reached_upload_mark = 0, m_reached_download_mark = 0;
-
-    void on_upload_completion();
+    version_type m_upload_completion_requested_version = -1;
+
     void on_download_completion();
```

Comment:
The end state of all this does need to be that download completion is also determinable by inspecting the Realm file, but it's significantly more complicated to get there for downloads (as the server is the source of truth for download completion rather than the client). I'm trying to split off each of the separate pieces to avoid having another monster PR that changes everything.